src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
     public void add(T value, int index) {
+        if (index != size) {
+            checkIndex(index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNode = head;
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            addInStart(newNode);
+            return;
+        } else {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            Node<T> prevNode = currentNode.prev;
+            newNode.next = currentNode;
+            newNode.prev = prevNode;
+            prevNode.next = newNode;
+            currentNode.prev = newNode;
+            incrementSize();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        Node<T> currentNode = head;
+        if (size == 1) {
+            newNode.prev = null;
+            newNode.next = null;
+            head = newNode;
+            tail = newNode;
+            return currentNode.value;
+        }
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T lastValue = currentNode.value;
+        currentNode.value = value;
+        return lastValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        if (index == 0) {
+            removeFromStart();
+            return currentNode.value;
+        }
+        if (index == size - 1) {
+            removeFromEnd();
+            return currentNode.value;
+        }
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object
+                    || (currentNode.value != null && currentNode.value.equals(object))) {
+                if (i == 0) {
+                    removeFromStart();
+                    return true;
+                }
+                if (i == size - 1) {
+                    removeFromEnd();
+                    return true;
+                }
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
```

Comment:
you can rewrite this code in such a way that won't need to handle edge cases in separate methods. Just iterate all the way from head or tail till you find that object, then `unlink` it and exit while loop returning `true`